### ISSUE TITLE
Cannot login using signing subkey without master private key

### ISSUE THREAD

**Issue by vemek** [keybase/keybase-issues#1968]
Hi,

I'm trying to login again now that keybase is using per-device keys, and I'm seeing the following error:
```
Sorry, you have a PGP private key that Keybase doesn't have a copy of.
You need to prove you're you. Any of these will unlock this computer:

   - reset your account and start fresh: https://keybase.io/#account-reset
   - install GPG and put your PGP private key on this machine and try again
   - go back and provision with another device or paper key
```

I have an offline master key, and use three (SEC) subkeys on my device. My assumption is that I should be able to use these to sign the device key, but keybase doesn't support that yet. Is this something you might be able to look at in future, or is my assumption wrong and I'll be required to use the master private key?

Details:
```
≡ keybase --version
keybase version 1.0.8-0
≡ gpg2 --version
gpg (GnuPG/MacGPG2) 2.0.28
libgcrypt 1.6.3
Copyright (C) 2015 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Home: ~/.gnupg
Supported algorithms:
Pubkey: RSA, RSA, RSA, ELG, DSA
Cipher: IDEA, 3DES, CAST5, BLOWFISH, AES, AES192, AES256, TWOFISH,
        CAMELLIA128, CAMELLIA192, CAMELLIA256
Hash: MD5, SHA1, RIPEMD160, SHA256, SHA384, SHA512, SHA224
Compression: Uncompressed, ZIP, ZLIB, BZIP2
≡ gpg2 -K
/Users/vemek/.gnupg/secring.gpg
-------------------------------
sec#  3744R/4733AC75 2015-09-24 [expires: 2017-09-23]
uid                  Mark King <xxxxx@xxx.com>
uid                  Mark King <xxxxx@xxx.com>
uid                  Mark King <xxxxx@xxx.com>
ssb>  2048R/3E3106C2 2015-09-24
ssb>  2048R/42445BC1 2015-09-24
ssb>  2048R/1058D21E 2015-09-24
```

Any and all help appreciated, thanks guys!

**Comment by maxtaco** [comment]
Can you post the output of 'keybase -d login'?

**Comment by brad2014** [comment]
Similar situation.  keybase does not recognize I my signing key is a subkey.

**Comment by maxtaco** [comment]
What does `gpg -K` say?

**Comment by brad2014** [comment]
With help from Max, I determined that my secring.gpg was corrupted (its primary key expiration date got out of sync with the expiration date for the same key in pubring.gpg) - fixing that resolved my problem.

**Comment by westonruter** [comment]
@brad2014 
> fixing that resolved my problem

how did you fix it?

**Comment by brad2014** [comment]
I deleted my public and private keys from their respective keyrings, and reimported them (as exported from my offline master, excluding the master key).  My guess is that the corruption occurred after changing the expiration date in the offline master, when I imported the updated public key and private subkeys *on top of* the existing keys - my guess is it was a gpg bug in merging the import into the existing keys. 

It could be related to GnuPG bug https://bugs.gnupg.org/gnupg/issue1817

**Comment by seefood** [comment]
Got the same error message, I'll try this expiration date trick.
but just to be sure, is this a bad phrasing or uhh... "Sorry, you have a PGP private key that Keybase doesn't have a copy of." - seems to suggest keybase expects to have a copy of my private key and I'm not quite sure it's supposed to.

**Comment by seefood** [comment]
nope, could not make the error go away. even tried exporting and importing in gpg2 just in case. also tried keybase --gpg /usr/bin/gpg2

Any smart advice? I might have an older copy of my secring, but it might not feature the keys I used with keybase. does that mean I have to reset my account?

Also, is there a reason Debian+Ubuntu treats gpg2 as a separate beast and not a newer version of the original tool, and can I set keybase to always use the new one? Is there a reason not to want that?

**Comment by maxtaco** [comment]
We will have to look at your bug separately ira.  It might be different.
Likely can't get to it right away.

**Comment by maxtaco** [comment]
@seefood can you post a transcript via `keybase -d login`? I can identify you properly, so that means it's a private and not public key issue

**Comment by seefood** [comment]
http://pastebin.com/yuMKS3xx
pasted the relevant part - starting with my passphrase getting approved. Should I open as a separate issue?

**Comment by maxtaco** [comment]
That transcript doesn't really help since it leaves out a lot of what I need to debug.  You can PGP email it to me if you want. There's nothing secret in there, by design, so people can post us debug logs.

**Comment by maxtaco** [comment]
@vemek looking more at your original problem, we do support this case and I just successfully tested it. There's something else going wrong in your example.  Happy to take a look for you.  Did you get the option to provision via `gpg`?  It might be a path problem that `keybase` can't find your `gpg` binary and therefore isn't going you the option.

**Comment by seefood** [comment]
http://pastebin.com/yuMKS3xx updated to a full dump. it does reveal my local username (via homedir path), custom search path and more :-)
So, worthy of a separate bug report?

**Comment by maxtaco** [comment]
Thank you! The problem is around line ~61 as far as I can tell.

Can you run this command:

```sh
/usr/bin/gpg2 --no-tty --with-colons --fingerprint -K
echo $?
```

Thanks!

**Comment by seefood** [comment]
Sure. I'm only breaking the Email addresses (though they are already all over spammers' lists)

❯ /usr/bin/gpg2 --no-tty --with-colons --fingerprint -K
gpg: Oops: keyid_from_fingerprint: no pubkey
gpg: Note: signatures using the MD5 algorithm are rejected
gpg: key AD29B051 marked as ultimately trusted
sec::14144:1:226142E355CF75A7:1242885979::::::::::
fpr:::::::::A8B347331CBC5EA93A009CC6226142E355CF75A7:
uid:::::::9A116947F6C9C2228066436FF923B70C67FA7915::Ira Abramov (Free Tinkerer) <pgpkey2009(a)ira.abramov.org>:
sec::10016:1:D56E1B4C135D47A1:1242948710::::::::::
fpr:::::::::C4AE6397FABF8F2B913EB0D5D56E1B4C135D47A1:
uid:::::::9A116947F6C9C2228066436FF923B70C67FA7915::Ira Abramov (Free Tinkerer) <pgpkey2009(a)ira.abramov.org>:
ssb::8192:1:D2BA914F1F936054:1243006915::::::::::
ssb::8896:1:150509B8F5FEEBA3:1243007302::::::::::
ssb::11136:1:A4F4A272575EFEF7:1243007632::::::::::
ssb::12224:1:62181DFB55B133B2:1243021790::::::::::
sec::4096:1:9439E86389D0AF41:1377776735:1630064735:::::::::
fpr:::::::::53D2A6E10BB70B384E52431D9439E86389D0AF41:
uid:::::::00C5B00F8769E17D05D2750FF8CB4DBE95E6120D::Ira Abramov (Email/Insecure key) <GnuPGmail-2013(a)ira.abramov.org>:
ssb::4096:1:A3DFAC0198D76C55:1377776735::::::::::
sec::1024:1:3D350C09AD29B051:783426257::::::::::
fpr:::::::::8CA21FA5BC29E1636CB0610E55AF0099:
uid:::::::82473495C5648F5758BDD7EF86D2EE21B34CB8CA::Ira Abramov <ira(a)scso.com>:
uid:::::::82473495C5648F5758BDD7EF86D2EE21B34CB8CA::Ira Abramov <ira(a)scso.com>:
uid:::::::440F8056BCA50E67DF3D8ED55C21A158CC1A039A::Ira Abramov <ia42(a)keybase.io>:
ssb::3072:16:780B9288177B9FE9:1009921571::::::::::
sec::1024:17:31CC32CEF78F3EE4:1009875755::::::::::
fpr:::::::::5A165A4F0C7BB6666838809F31CC32CEF78F3EE4:
uid:::::::0B82095F761A6F634F33AE94E1648FB504B350AF::Ira Abramov <pgpkey-20020101(a)ira.abramov.org>:
uid:::::::440F8056BCA50E67DF3D8ED55C21A158CC1A039A::Ira Abramov <ia42(a)keybase.io>:
ssb::3072:16:6AB8E65E3842A813:1009875892::::::::::
exit is 2

**Comment by maxtaco** [comment]
Hm, that helps! One much thing to check out: 

```sh
/usr/bin/gpg2 --no-tty --with-colons --fingerprint -K > /dev/null
```

I'm curious to know if those warnings are shunted to stderr.  Then we can work around this issue.

**Comment by seefood** [comment]
❯ /usr/bin/gpg2 --no-tty --with-colons --fingerprint -K > /dev/null
gpg: Oops: keyid_from_fingerprint: no pubkey
gpg: Note: signatures using the MD5 algorithm are rejected
gpg: key AD29B051 marked as ultimately trusted

I have noticed this before, how only one of my 4 secret keys is "ultimately trusted", have no clue why it is different. my config file lists:

default-key 31CC32CEF78F3EE4  (That's the one known by keybase)
trusted-key 31CC32CEF78F3EE4
trusted-key 3D350C09AD29B051

**Comment by maxtaco** [comment]
Those three statements from gpg might not be related.  The first one seems
like the real issue though I have never seen it before.  But that is
causing gpg to return non-0 which is stopping the import.

**Comment by maxtaco** [comment]
Check out this related issue BTW: https://github.com/keybase/keybase-issues/issues/821
